### ISSUE TITLE
Caching offer sdp when master is not available yet

### ISSUE THREAD

**Issue by zwei333** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#488]
In our system, app could send offer to camera before camera's webrtc is ready.
It will be convenient that signaling channel caches the offer sdp message, and sends it to camera (master) when webrtc is ready.  In this way, app doesn't have to resend the message.
Does kvs code support this feature?  If not, can it be added to the SDK?
My initial test shows negative.  I first start sample JS webApp streaming, which sends out offer.  Then I starte camera.  sample JS webApp  never receives answer without sending another offer.

**Comment by MushMal** [comment]
@zwei333 the SDK is a static library that gets linked against an application. If the application is not running there is no way to pre-cache the offer. 

This is the reason why we implement "delivery' mechanism on the backend for the signaling channel. When creating a signaling channel, you specify message TTL. This is basically how long the message can stay up in the backend before it gets dropped or delivered.

Say, you've configured your channel with message TTL of 10 minutes. The viewer can send an offer to the master that hasn't yet started. This is OK. The master application starts later and connects to the signaling channel before 10 minutes expire. The message will be delivered to the master in this case. If, however, the master connects after the message TTL expires the message will be lost and the master won't receive the offer.

Please resolve this if you don't have any further questions.

**Comment by zwei333** [comment]
When I use the word "cache", actually I meant we expect the message is stored in the signaling channel backend.
But I don't see this deferred delivery.  Maybe the message TTL is set to zero.  
Who creates the signaling channel? the master, or viewer, or any one comes first?
We do not use key/secret, but IoT credentialing instead.  Some time ago I heard we are using pre-created signaling channel (I am not sure about it.  Correct me if I am wrong).
In our case, where should I check the TTL value?

**Comment by MushMal** [comment]
@zwei333 if you are using IoT then most likely your channel has been pre-created. Irregardless of the integration, the default behavior of the signaling client state machine is to issue a Describe call followed by Create call if the channel doesn't exist. However, if the channel exists, it skips Create and moves to the GetEndpoint state. If the signaling client needs to create the channel, it will use the TTL from the ChannelInfo structure supplied by the application.

Please check the channel's TTL. Check your application and ensure you have the TTL set properly. 

NOTE: there is a min TTL which is defined as 5 seconds and  max TTL which value which is 120 seconds: https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_SingleMasterConfiguration.html#KinesisVideo-Type-SingleMasterConfiguration-MessageTtlSeconds 

default is 0 which maps to 60 seconds

**Comment by zwei333** [comment]
Tried to send offer before camera's webrtc starts.  webrtc didn't receive the offer after start.
Have seen this issue with key/secret credentialing and connecting from KVS sample JS app.
Also got the same result with IoT credentialing and connecting from our app.
For this issue, what kind of tool is available for debugging?

**Comment by MushMal** [comment]
Could you please email your account id to Kinesis-video-support@amazon.com and include your channel name and if you can the request id?

If I were to guess, there is an issue with a timeout or something similar

**Comment by bdhandap** [comment]
@zwei333 Viewer can connect before the master and when the master connects it should receive any pending SDP message (offer/answer/candidate) if it is within TTL. However, there is an intermittent bug in Signaling service where the SDP message (offer/answer/ice-candidate) is not delivered during connect. We do have a fix in place which is pending deployment. I will update this thread once it is deployed.

**Comment by zwei333** [comment]
Do we have an expected time of the fix deployment?

**Comment by bdhandap** [comment]
@zwei333 Changes are deployed. It should work now. Can you try again ?

**Comment by zwei333** [comment]
Have tried, but no change.  Started viewer first and camera later.  offer is lost

**Comment by bdhandap** [comment]
@zwei333 Thanks for trying. I am checking the logs and will let you know.

**Comment by zwei333** [comment]
Can we have a short debug meeting to go through this bug?

**Comment by MushMal** [comment]
Can someone update this thread? What's the latest on this issue?

**Comment by bdhandap** [comment]
@zwei333 Found the actual root cause for this issue and the fix has been deployed to eu-central-1. Other regions will be deployed shortly. Can you verify whether you are able to start the viewer first in eu-central-1 and see if it works?

I verified in eu-central-1 using JS Sample https://awslabs.github.io/amazon-kinesis-video-streams-webrtc-sdk-js/examples/index.html

**Comment by zwei333** [comment]
I tried eu-central-1. Our webrtc behaves differently.  I think your theory above is correct.
However, the cached offer message caused our webrtc to crash.  I guess the crash is caused by early delivery of offer.  Therefore, I can't stream with current version of webrtc.  I will fix the crash and try again.
Please go ahead with deploying the fix

**Comment by zwei333** [comment]
I fixed our webrtc crash issue in this scenario. 
Now, using eu-central-1, I can start viewer first, and launch master (webrtc) later.  Once master is started, cached messages are delivered and video starts shortly after.
Waiting for the fix to be deployed to us-east-1

**Comment by zwei333** [comment]
Tested with us-east-1.  Working now.  Closing this issue

**Comment by zwei333** [comment]
Need to re-open this issue.  It does not work all the time.
After a camera power cycle, if viewer sends offer to signaling server before running webrtc master on camera, when webrtc master starts for the first time, the offer will be lost, so viewer can't connect to master.  But repeating the same operation without a power cycle will result in successful connection.
Steps to reproduce (using https://awslabs.github.io/amazon-kinesis-video-streams-webrtc-sdk-js/examples/index.html as viewer):
0. run webrtc and then viewer, stream OK
1. power down camera
2. start viewer and then power on camera ( Or, power on camera and start viewer, gets the same result)
3. start webrtc the first time.  offer is lost.  Can't stream
4. stop viewer and kill webrtc
5. start viewer
6. start webrtc the second time.  offer is received, can stream.

Did a lot of investigation.  The same problem is observed on two different platform.  So, it is not related to platform drivers and network stack.
Then discovered something with tcpdump
Steps:
0. run webrtc and then viewer, stream OK
1. power down camera and then power on camera
2. start tcpdump ( tcpdump port 443)
3. start viewer.  Within one second, tcpdump shows that signaling channel sends a few packets to camera, using the old port number it used to talk to webrtc BEFORE the power cycle
4. start webrtc.  offer lost.  no streaming.

I think there is a bug in signaling channel caching. After the power cycle, signaling channel sends the cached offer to the old port number used before power cycle even when webrtc is not running, and thus  the offer is lost.  Please note, we run webrtc always with the same channel name.

To prove it, I made this work around to get streaming running OK
Steps:
0.  run webrtc and then viewer, stream OK
**<WORK_AROUND>  Before power down, kill webrtc, stop viewer. And then start viewer, run for 2 seconds.  This extra run of viewer will correct signaling channel caching (Why? I don't know)**
1. power down camera and then power on camera
2. start tcpdump ( tcpdump port 443)
3. start viewer.  Wait for a few seconds, no packets from signaling channel are sent to camera
4. start webrtc.  offer is received.  streaming fine.

I repeated the above expriement many times on both platforms.  100% reproducible.

**Comment by MushMal** [comment]
@zwei333 unfortunately, we will not be able to debug your application on your platform. The powercycle is irrelevant to the signaling. 

Try to use an open platform such as Linux or Mac. Run the stock master application without any modifications. Ensure your signaling channel is created with enough TTL - for example 10 minutes. This will guarantee that when the viewer sends the message, the master will have up-to 10 minutes to fetch the message.

Once this setup is done, try to reproduce it (no need for restarting of your computer). If you are able to reproduce this then capture the debug logs from the master including the channel name and account ID so we can dive into it.

This really is not actionable from our side. You might want to check your hardware or the changes on your software. Try to capture debug logs which might give you some ideas on what to do.

**Comment by zwei333** [comment]
To be clear, there is no problem with webrtc master code.
After power cycle, by running https://awslabs.github.io/amazon-kinesis-video-streams-webrtc-sdk-js/examples/index.html sample viewer, tcpdump shows that signaling  server sends packets to camera even when webrtc is not running.
In the work around, I designed an experiment so that signaling server doesn't send packets to camera when webrtc is not running. Then when webrtc starts, the offer is received properly.   It can prove that the unexpected packets from signaling server contained the offer and was lost.
The master log will not help.  The problem happens before master starts

**Comment by MushMal** [comment]
I am puzzled really. The offer is exchanged via websocket so there must be a handshake - a connect event. 
What I suspect is that the previous session was not gracefully terminated and the backend still thinks the connection is still open so it sends the packets. The client , just like the backend is sending pongs at regular intervals to indicate that it’s still up. If the other side is gone without termination the websocket will not know that it’s gone until the pong fails. 


Could you do the same experiment but with the following... connect viewer and master. Power down the master and wait for say 10 seconds. Use the viewer to send the offer. Start the master. This time, the offer should arrive ok.

**Comment by zwei333** [comment]
Yes, the offer should be sent after websocket connection is created.  But I do see that after camera is started and before webrtc is running, by starting viewer, a few packets are sent to camera using the old port number.  Whenever this happens, the offer never makes it into webrtc when webrtc is started later.  Below is tcpdump log:


0. run webrtc and then viewer, stream OK.  Stop viewer

2020-09-22 14:06:33 DEBUG   lwsWssCallbackRoutine(): Client is writable
22:06:33.219997 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [P.], seq 24212:24243, ack 4505, win 160, options [nop,nop,TS val 708663847 ecr 42949624361
22:06:33.220112 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24243, win 2641, options [nop,nop,TS val 4294962439 ecr 708663847], length 0
22:06:36.221601 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24243, win 2641, options [nop,nop,TS val 4294962740 ecr 708663847], length 0
22:06:36.254811 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4505, win 160, options [nop,nop,TS val 708664606 ecr 4294962439], length 0
22:06:39.251642 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24243, win 2641, options [nop,nop,TS val 4294963043 ecr 708664606], length 0
22:06:39.290051 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4505, win 160, options [nop,nop,TS val 708665364 ecr 4294962439], length 0
22:06:42.281614 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24243, win 2641, options [nop,nop,TS val 4294963346 ecr 708665364], length 0
22:06:42.311238 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4505, win 160, options [nop,nop,TS val 708666120 ecr 4294962439], length 0
22:06:43.220912 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [P.], seq 4505:4540, ack 24243, win 2641, options [nop,nop,TS val 4294963439 ecr 7086661205
2020-09-22 14:06:43 DEBUG   lwsWssCallbackRoutine(): Client is writable
22:06:43.252698 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [P.], seq 24243:24274, ack 4540, win 160, options [nop,nop,TS val 708666355 ecr 42949634391
22:06:43.252891 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24274, win 2641, options [nop,nop,TS val 4294963443 ecr 708666355], length 0
22:06:46.261634 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24274, win 2641, options [nop,nop,TS val 4294963744 ecr 708666355], length 0
22:06:46.296520 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4540, win 160, options [nop,nop,TS val 708667116 ecr 4294963443], length 0
22:06:49.291603 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24274, win 2641, options [nop,nop,TS val 4294964047 ecr 708667116], length 0
22:06:49.323371 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4540, win 160, options [nop,nop,TS val 708667873 ecr 4294963443], length 0
22:06:52.321626 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24274, win 2641, options [nop,nop,TS val 4294964350 ecr 708667873], length 0
22:06:52.357891 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4540, win 160, options [nop,nop,TS val 708668631 ecr 4294963443], length 0
22:06:53.257671 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [P.], seq 4540:4575, ack 24274, win 2641, options [nop,nop,TS val 4294964443 ecr 7086686315
2020-09-22 14:06:53 DEBUG   lwsWssCallbackRoutine(): Client is writable
22:06:53.290464 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [P.], seq 24274:24305, ack 4575, win 160, options [nop,nop,TS val 708668865 ecr 42949644431
22:06:53.290581 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24305, win 2641, options [nop,nop,TS val 4294964446 ecr 708668865], length 0
22:06:56.291602 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24305, win 2641, options [nop,nop,TS val 4294964747 ecr 708668865], length 0
22:06:56.326233 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4575, win 160, options [nop,nop,TS val 708669623 ecr 4294964446], length 0
22:06:59.321658 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [.], ack 24305, win 2641, options [nop,nop,TS val 4294965050 ecr 708669623], length 0
22:06:59.357576 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], ack 4575, win 160, options [nop,nop,TS val 708670381 ecr 4294964446], length 0

1. power down camera, wait for 1 minute, and then power on camera
2. start tcpdump ( tcpdump port 443)
3. start viewer. webrtc is not running yet.

22:08:25.633449 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], seq 379741590:379743038, ack 3812044538, win 160, options [nop,nop,TS val 7086919468
22:08:25.633573 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [R], seq 3812044538, win 0, length 0
22:08:25.636453 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], seq 1448:2896, ack 1, win 160, options [nop,nop,TS val 708691946 ecr 4294964446], len8
22:08:25.636566 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [R], seq 3812044538, win 0, length 0
22:08:25.638917 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], seq 2896:4344, ack 1, win 160, options [nop,nop,TS val 708691946 ecr 4294964446], len8
22:08:25.639042 IP 42498ccf651e5f0d5050e20fc0017e5f.46985 > ec2-52-72-115-243.compute-1.amazonaws.com.443: Flags [R], seq 3812044538, win 0, length 0
22:08:25.641349 IP ec2-52-72-115-243.compute-1.amazonaws.com.443 > 42498ccf651e5f0d5050e20fc0017e5f.46985: Flags [.], seq 4344:5792, ack 1, win 160, options [nop,nop,TS val 708691946 ecr 4294964446], len8

In these four messages, port 46985 is the same as before power cycle.  But seq numbers are garbage.

**Comment by zwei333** [comment]
Any update on this issue?

**Comment by lherman-cs** [comment]
tagging @bdhandap and @sharadrj

**Comment by zwei333** [comment]
A touch to keep this issue alive

**Comment by zwei333** [comment]
No new update. If power is pulled from camera this problem is 100% re-producible.

**Comment by MushMal** [comment]
We can't really reproduce this as it might be platform specific. Are you able to see this issue reproduced when the application is restarted or any other way we can try to replicate

**Comment by zwei333** [comment]
@bkneff  and I had discussed this issue over the phone a few times.  Please ask him for details

**Comment by docee** [comment]
We have a similar problem.  The probability of this problem in ap-northeast-1 is very high.  When the Master connects successfully, the Master cannot receive any offers or candidates.

**Comment by MushMal** [comment]
@docee the issue described here is very specific to a device reboot scenario where the message is "being delivered" to a non-existent connection. Could you please clarify your scenario? Do you have this happening when your device reboots?

**Comment by docee** [comment]
> @docee the issue described here is very specific to a device reboot scenario where the message is "being delivered" to a non-existent connection. Could you please clarify your scenario? Do you have this happening when your device reboots?

YES. It may also happened after the device restarts. So I still can't tell whether it is the SDK problem or the server problem.

**Comment by MushMal** [comment]
@docee could you please get more data and see if it’s related to this issue. As far as we are aware there are no other scenarios which lead to the messages are not being delivered.

**Comment by docee** [comment]
> @docee could you please get more data and see if it’s related to this issue. As far as we are aware there are no other scenarios which lead to the messages are not being delivered.

I have tested cn-north-1, ap-northeast-1, us-east-1, us-west-2, and currently only cn-north-1 and us-east-1 have this problem. So I am not sure if it is the server problem.

Sometimes this problem occurs when the signaling client is created for the first time after the device is started. But sometimes this problem occurs when a new signaling client is re-established after the signaling is disconnected.

When this happens, RECONNECT can solve the problem.

The following is my log. KVS is not outputting any logs from the last line. I only record the INFO level log of KVS, so it has no reference significance.
```
[main] 1605622834588 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> New
[main] 1605622834589 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Get Security Credentials
[main] 1605622834591 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Get Channel Endpoint
[main] 1605622834594 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Get ICE Server Configuration
[main] 1605622834594 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Ready
[main] 1605622834594 I src/conn/src/kvs_webrtc/kvs.c(122):[signaling-2] - SignalingClient signalingClientConnectSync(): Signaling Client Connect Sync
[main] 1605622834595 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Connecting
[main] 1605622835191 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Get ICE Server Configuration
[main] 1605622835192 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Ready
[main] 1605622835192 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Connecting
[main] 1605622835193 I src/conn/src/kvs_webrtc/kvs_signaling.c(177):[signaling-2] - state -> Connected
[main] 1605622835193 I src/conn/src/kvs_webrtc/kvs.c(122):[signaling-2] - SignalingClient signalingClientGetIceConfigInfoCount(): Signaling Client Get ICE Config Info Count
[main] 1605622835193 I src/conn/src/kvs_webrtc/kvs_utils.c(39):[signaling-2] - ice config info count 1
[main] 1605622835194 I src/conn/src/kvs_webrtc/kvs.c(122):[signaling-2] - SignalingClient signalingClientGetIceConfigInfo(): Signaling Client Get ICE Config Info
[main] 1605622835194 D src/conn/src/kvs_webrtc/kvs_rtc_connection_factory.c(21):[signaling-2] - iceServer[#00]  url:stun:stun.kinesisvideo.cn-north-1.amazonaws.com.cn:443  username:  credential:
[main] 1605622835194 D src/conn/src/kvs_webrtc/kvs_rtc_connection_factory.c(21):[signaling-2] - iceServer[#01]  url:turn:54-223-38-244.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=udp  username:1605623133:djE6YXJuOmF3cy1jbjpraW5lc2lzdmlkZW86Y24tbm9ydGgtMTowMzEyNTY2Mzg3NjU6Y2hhbm5lbC80MDM2NTAyNDUwMTAwMDAxLzE2MDU2MTM0NzE3OTU=  credential:8pCAgfessZ9HV1FO8Xvmpp19SrHfSHgiQq9z55Axkq4=
[main] 1605622835194 D src/conn/src/kvs_webrtc/kvs_rtc_connection_factory.c(21):[signaling-2] - iceServer[#02]  url:turns:54-223-38-244.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=udp  username:1605623133:djE6YXJuOmF3cy1jbjpraW5lc2lzdmlkZW86Y24tbm9ydGgtMTowMzEyNTY2Mzg3NjU6Y2hhbm5lbC80MDM2NTAyNDUwMTAwMDAxLzE2MDU2MTM0NzE3OTU=  credential:8pCAgfessZ9HV1FO8Xvmpp19SrHfSHgiQq9z55Axkq4=
[main] 1605622835194 D src/conn/src/kvs_webrtc/kvs_rtc_connection_factory.c(21):[signaling-2] - iceServer[#03]  url:turns:54-223-38-244.t-89e91582.kinesisvideo.cn-north-1.amazonaws.com.cn:443?transport=tcp  username:1605623133:djE6YXJuOmF3cy1jbjpraW5lc2lzdmlkZW86Y24tbm9ydGgtMTowMzEyNTY2Mzg3NjU6Y2hhbm5lbC80MDM2NTAyNDUwMTAwMDAxLzE2MDU2MTM0NzE3OTU=  credential:8pCAgfessZ9HV1FO8Xvmpp19SrHfSHgiQq9z55Axkq4=
```
